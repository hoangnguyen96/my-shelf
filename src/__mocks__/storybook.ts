export const mockRouter = {
  pathname: "/",
  route: "/",
  query: {},
  asPath: "/",
  basePath: "",
  push: () => Promise.resolve(true),
  replace: () => Promise.resolve(true),
  reload: () => {},
  back: () => {},
  prefetch: () => Promise.resolve(),
  beforePopState: () => {},
  isFallback: false,
  isLocaleDomain: false,
  isReady: true,
  isPreview: false,
  forward: () => {},
  events: {
    on: () => {},
    off: () => {},
    emit: () => {},
  },
};
