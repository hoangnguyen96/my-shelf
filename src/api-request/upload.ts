import { MESSAGES } from "@app/constants";

export const generateImageUpload = async (file: File) => {
  try {
    const formData = new FormData();
    formData.append("image", file);

    const response = await fetch(
      `https://api.imgbb.com/1/upload?key=${process.env.NEXT_PUBLIC_IMGBB_API_KEY}`,
      {
        method: "POST",
        body: formData,
      }
    );

    if (!response.ok) {
      throw new Error(MESSAGES.UPLOAD_FAILED);
    }

    const data = await response.json();
    return data;
  } catch (error) {
    throw new Error(MESSAGES.ERROR_UPLOAD);
  }
};
