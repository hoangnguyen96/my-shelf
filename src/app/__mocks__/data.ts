export const DATA_BOOKS = [
  {
    title: "Don't Make Me Think",
    author: "Steve Krug",
    category: "SelfHelpBook",
    imageUrl: "https://i.ibb.co/QbQgVtG/book1.png",
    description: "description will update later",
    status: false,
    publicationYear: 2000,
    rating: 4.5,
    createdDate: "3 Sep 2024 9:07 PM",
    edition: "first",
    id: "1",
  },
  {
    title: "Rich Dad Poor Dad",
    author: "Robert T. Kiyosaki",
    category: "SelfHelpBook",
    imageUrl: "https://i.ibb.co/9qHTtQF/book2.png",
    description:
      "Steve Krug is a usability consultant who has more than 30 years of experience as a user advocate for companies like Apple, Netscape, AOL, Lexus, and others. Based in part on the success of his first book, Don't Make Me Think, he has become a highly sought-after speaker on usability design.",
    status: false,
    publicationYear: 1997,
    rating: 4.5,
    createdDate: "3 Sep 2024 8:47 PM",
    edition: "second",
    id: "2",
  },
  {
    title: "The Design of Everyday Things",
    author: "Don Norman",
    category: "TextBook",
    imageUrl: "https://i.ibb.co/LnSBcCM/book5.png",
    description:
      "Steve Krug is a usability consultant who has more than 30 years of experience as a user advocate for companies like Apple, Netscape, AOL, Lexus, and others. Based in part on the success of his first book, Don't Make Me Think, he has become a highly sought-after speaker on usability design.",
    status: false,
    publicationYear: 1988,
    rating: 4.5,
    createdDate: "3 Sep 2024 8:51 PM",
    edition: "fourth",
    id: "3",
  },
];

export const DATA_USER = [
  {
    username: "admin",
    email: "admin@gmail.com",
    password: "$2a$10$MwSZ9vWSX7DVJwJ4SF5Mzuj2zMFjY52XM9L11gGG0j79gybowaLfq",
    isAdmin: true,
    phone: "",
    bio: "",
    avatar: "https://i.ibb.co/RHMqQGr/man-1.png",
    favorites: ["3", "2", "1", "13", "5"],
    shelfBooks: ["1", "6"],
    userId: "3733403",
    id: "1",
  },
  {
    username: "hung",
    email: "hung1@gmail.com",
    password: "$2a$10$Mt6ruwOzt3nzX1.HpSt.OOzut.teXmLE74WclmkP.Nn.gzVQrlqri",
    isAdmin: false,
    phone: "",
    bio: "",
    avatar: "https://i.ibb.co/88X1WfZ/avatar-default.png",
    favorites: ["1"],
    shelfBooks: ["2", "3", "1"],
    userId: "8857134",
    id: "2",
  },
];
