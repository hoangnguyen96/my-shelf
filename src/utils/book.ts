import { DEFAULT_LIMIT, TYPE_SEARCH } from "@app/constants";
import { BookType } from "@app/interface";

export const filterBooksOnShelf = (
  allBooks: BookType[],
  shelfBooks: string[]
) => allBooks.filter((item) => shelfBooks.includes(item.id));

export const filterBooksOnShelfByParams = (
  booksOnShelf: BookType[],
  type: string,
  value: string
) =>
  booksOnShelf.filter((item) =>
    type === TYPE_SEARCH.TITLE && value
      ? item.title.toLowerCase().includes(value.toLowerCase())
      : type === TYPE_SEARCH.AUTHOR && value
        ? item.author.toLowerCase().includes(value.toLowerCase())
        : item
  );

export const filterBooksFavorite = (
  allBooks: BookType[],
  favorites: string[]
) => allBooks.filter((item) => favorites.includes(item.id));

export const filterBooksFavoriteByParams = (
  booksFavorites: BookType[],
  type: string,
  value: string
) =>
  booksFavorites.filter((item) =>
    type === TYPE_SEARCH.TITLE && value
      ? item.title.toLowerCase().includes(value.toLowerCase())
      : type === TYPE_SEARCH.AUTHOR && value
        ? item.author.toLowerCase().includes(value.toLowerCase())
        : item
  );

export const dividePaginationBooks = (list: BookType[], limit?: number) => {
  return list.reduce((acc: BookType[][], _, i, self) => {
    if (!(i % (limit || DEFAULT_LIMIT))) {
      return [...acc, self.slice(i, i + (limit || DEFAULT_LIMIT))];
    }
    return acc;
  }, []);
};
