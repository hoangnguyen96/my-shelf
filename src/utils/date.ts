import { MONTHS, TIME_PERIODS } from "@app/constants";

export const formatDate = (date: Date) => {
  const hours = date.getHours();
  const minutes = date.getMinutes();
  const period = hours >= 12 ? TIME_PERIODS.PM : TIME_PERIODS.AM;
  const formattedHours = hours % 12 || 12;
  const formattedMinutes = minutes < 10 ? `0${minutes}` : minutes;

  return `${date.getDate()} ${MONTHS[date.getMonth()]} ${date.getFullYear()} ${formattedHours}:${formattedMinutes} ${period}`;
};
