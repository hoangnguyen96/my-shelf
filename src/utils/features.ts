import { DEFAULT_LIMIT, TYPE_SEARCH } from "@app/constants";
import { BookType } from "@app/models";
import {
  FieldErrors,
  FieldValues,
  Path,
  UseFormClearErrors,
} from "react-hook-form";
import { v4 as uuidv4 } from "uuid";

/**
 * Clear error message when the users typing
 */
export const clearErrorOnChange = <T extends FieldValues>(
  fieldName: Path<T>,
  errors: FieldErrors<T>,
  clearErrorFunc: UseFormClearErrors<T>
): void => {
  errors[fieldName]?.message && clearErrorFunc(fieldName);
};

export const generateSevenDigitUUID = () => {
  const uuid = uuidv4().replace(/\D/g, "");
  return uuid.slice(0, 7);
};

export const getTopBook = (data: BookType[], limit?: number) =>
  data
    .sort(
      (a, b) =>
        new Date(b.createdDate).getTime() - new Date(a.createdDate).getTime()
    )
    .slice(0, limit || DEFAULT_LIMIT);

export const formatDate = (date: Date) => {
  const months = [
    "Jan",
    "Feb",
    "Mar",
    "Apr",
    "May",
    "Jun",
    "Jul",
    "Aug",
    "Sep",
    "Oct",
    "Nov",
    "Dec",
  ];

  const hours = date.getHours();
  const minutes = date.getMinutes();
  const period = hours >= 12 ? "PM" : "AM";
  const formattedHours = hours % 12 || 12;
  const formattedMinutes = minutes < 10 ? `0${minutes}` : minutes;

  return `${date.getDate()} ${months[date.getMonth()]} ${date.getFullYear()} ${formattedHours}:${formattedMinutes} ${period}`;
};

export const dividePaginationBooks = (list: BookType[]) => {
  return list.reduce((acc: BookType[][], _, i, self) => {
    if (!(i % 12)) {
      return [...acc, self.slice(i, i + 12)];
    }
    return acc;
  }, []);
};

export const getFirstPath = (value: string) => {
  if (!value) {
    return "";
  }

  const parts = value.split("/");
  const basePath = `/${parts[1]}`;

  return basePath;
};

export const getTwoPath = (value: string) => {
  if (!value) {
    return "";
  }

  const parts = value.split("/");
  const basePath = `/${parts[1]}/${parts[2]}`;

  return basePath;
};

export const filterBooksOnShelf = (
  allBooks: BookType[],
  shelfBooks: string[]
) => allBooks.filter((item) => shelfBooks.includes(item.id));

export const filterBooksOnShelfByParams = (
  booksOnShelf: BookType[],
  type: string,
  value: string
) =>
  booksOnShelf.filter((item) =>
    type === TYPE_SEARCH.TITLE && value
      ? item.title.toLowerCase().includes(value.toLowerCase())
      : type === TYPE_SEARCH.AUTHOR && value
        ? item.author.toLowerCase().includes(value.toLowerCase())
        : item
  );

export const filterBooksFavorite = (
  allBooks: BookType[],
  favorites: string[]
) => allBooks.filter((item) => favorites.includes(item.id));

export const filterBooksFavoriteByParams = (
  booksFavorites: BookType[],
  type: string,
  value: string
) =>
  booksFavorites.filter((item) =>
    type === TYPE_SEARCH.TITLE && value
      ? item.title.toLowerCase().includes(value.toLowerCase())
      : type === TYPE_SEARCH.AUTHOR && value
        ? item.author.toLowerCase().includes(value.toLowerCase())
        : item
  );

export const getListDataByTypeAndValue = (
  type: string,
  value: string,
  dataByParams: BookType[][],
  dataAll: BookType[][],
  pagination: number
) =>
  type && value ? dataByParams[pagination] || [] : dataAll[pagination] || [];

export const getDataByParams = (
  type: string,
  value: string,
  dataByParams: BookType[][],
  dataAll: BookType[][]
) => (type && value ? dataByParams : dataAll);
