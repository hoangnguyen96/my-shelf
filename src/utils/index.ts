export * from "./book";
export * from "./date";
export * from "./error";
export * from "./search";
export * from "./user";
export * from "./validators";
