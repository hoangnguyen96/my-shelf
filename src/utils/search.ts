export const getFirstPath = (value: string) => {
  if (!value) {
    return "";
  }

  const parts = value.split("/");
  const basePath = `/${parts[1]}`;

  return basePath;
};
