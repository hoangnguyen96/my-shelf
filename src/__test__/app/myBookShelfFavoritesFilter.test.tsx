import { act, fireEvent, render } from "@testing-library/react";
import { useSession } from "next-auth/react";
import { getAllBook, getBookById, getUserById } from "@app/api-request";
import { DATA_BOOKS, DATA_USER } from "@app/__mocks__/data";
import * as utils from "@app/utils";
import MyBookShelfFavorites from "@app/app/(dashboard)/my-book-shelf/(main)/favorites/[...slug]/page";

jest.mock("@app/api-request", () => ({
  getAllBook: jest.fn(),
  getUserById: jest.fn(),
  getBookById: jest.fn(),
  updateUserById: jest.fn(),
}));

jest.mock("next-auth/react", () => ({
  useSession: jest.fn(),
}));

jest.mock("next/navigation", () => ({
  useRouter: jest.fn(),
}));

jest.mock("@app/utils", () => ({
  ...jest.requireActual("@app/utils"),
  filterBooksFavorite: jest.fn(),
  filterBooksFavoriteByParams: jest.fn(),
}));

jest.mock("@app/actions/auth", () => ({
  logout: jest.fn(),
}));

describe("My Book Shelf Favorites Search", () => {
  const mockBooks = DATA_BOOKS.filter((item) =>
    DATA_USER[0].favorites.includes(item.id)
  );

  const mockBooksByParams = mockBooks.filter((book) =>
    book.title.includes("on")
  );

  (useSession as jest.Mock).mockReturnValue({
    data: {
      user: {
        isAdmin: true,
        email: "admin@gmail.com",
        id: "3733403",
        name: "admin",
        image: "https://i.ibb.co/RHMqQGr/man-1.png",
      },
      expires: "2024-12-31T23:59:59.999Z",
    },
    status: "authenticated",
    signIn: jest.fn(),
    signOut: jest.fn(),
  });

  beforeEach(() => {
    jest.clearAllMocks();
    jest.spyOn(utils, "filterBooksFavorite").mockReturnValue(mockBooks);
    jest
      .spyOn(utils, "filterBooksFavoriteByParams")
      .mockReturnValue(mockBooksByParams);

    (getAllBook as jest.Mock).mockReturnValue({
      data: DATA_BOOKS,
    });
    (getUserById as jest.Mock).mockReturnValue({
      favorites: DATA_USER[0].favorites,
    });
    (getBookById as jest.Mock).mockReturnValue({
      data: mockBooks[0],
    });
  });

  it("Should render correctly snapshot", async () => {
    await act(async () => {
      const { asFragment } = render(
        <MyBookShelfFavorites params={{ slug: ["title", "on"] }} />
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });

  it("Should handle fetch data failed", async () => {
    (getAllBook as jest.Mock).mockRejectedValue(
      new Error("Failed to fetch books")
    );
    await act(async () => {
      const { asFragment } = render(
        <MyBookShelfFavorites params={{ slug: ["title", "on"] }} />
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });

  it("Should handle favorite book", async () => {
    const { findAllByTestId } = render(
      <MyBookShelfFavorites params={{ slug: ["title", "on"] }} />
    );

    const buttons = await findAllByTestId("handle-favorite");

    fireEvent.click(buttons[0]);
  });
});
