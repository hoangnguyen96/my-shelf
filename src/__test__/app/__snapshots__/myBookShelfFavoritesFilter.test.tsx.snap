// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`My Book Shelf Favorites Search Should handle fetch data failed 1`] = `<DocumentFragment />`;

exports[`My Book Shelf Favorites Search Should render correctly snapshot 1`] = `<DocumentFragment />`;
